### ISSUE TITLE
Editor sets a linked mesh as the parent of another linked mesh when duplicating a skin

### ISSUE THREAD

**Issue by HaraldCsaszar** [EsotericSoftware/spine-runtimes#1284]
As reported on the forum: http://esotericsoftware.com/forum/Error-reading-Skeleton-JSON-file-in-Unity-11567?p=51910#p51910

Reproduction assets sent in this email:
FMfcgxwBVqSSwKQhdSJjxRDWcKzjfhxK

The following line tests for `!= null` but then still dereferences regionUVs.length:
https://github.com/EsotericSoftware/spine-runtimes/blob/3.7/spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/attachments/MeshAttachment.java#L75

[Edit:]
Problem seems to be multiple indirections of LinkedMesh in combinarion with unsorted order of the list of LinkedMeshes. Therefore the second indirection resolves to read values of the first indirection - however they are not yet filled out, but would be filled out later while iterating through the list of LinkedMeshes.

Details:
Error occurs in the reproduction project at the 8th entry of the LinkedMesh list "CH_FoG/Hairs_Patch" which has two indirections, at least when running the importer in Unity.
```
---

- [ ] spine-libgdx

---
- [ ] spine-csharp
- [ ] spine-xna
- [ ] spine-monogame
- [ ] spine-unity
- [ ] spine-tk2d

---
- [ ] spine-c
- [ ] spine-cpp
- [ ] spine-sfml
- [ ] spine-cocos2d-objc
- [ ] spine-cocos2dx
- [ ] spine-ue4

---
- [ ] spine-ts core
- [ ] spine-ts webgl
- [ ] spine-ts canvas
- [ ] spine-ts threejs

---
- [ ] spine-as3
- [ ] spine-starling

---
- [ ] spine-lua 
- [ ] spine-love
- [ ] spine-corona

**Comment by badlogic** [comment]
The code in the implementation is correct. The `||` is a short circuiting operator. If the null check succeeds, the rest of the expression is not evaluate, so there can not be an NPE.

The attachment in question is a linked mesh, meaning it gets its vertices (including uvs) from another mesh.

**Comment by HaraldCsaszar** [comment]
It is not the first `regionUVs.length` in the if clause that is the problem, but the second one on the same line that is causing the trouble because of the if condition evaluating to true.

**Comment by HaraldCsaszar** [comment]
For reference - this forum post mentions some additional information:
http://esotericsoftware.com/forum/Animations-synchro-in-the-previews-and-in-the-engine-11563

**Comment by badlogic** [comment]
The linked meshes require a topological sort, with the assumption that there are no cycles only containing linked meshes (meaning at least one linked mesh in a cycle must reference a non-linked mesh). Uses a scratch buffer to keep garbage and insertion/removal time complexity on the low side. Needs to be ported to SkeletonBinary as well. Works for the user project. https://github.com/EsotericSoftware/spine-runtimes/commit/85e54942f75c6ac9fe7099df4bcaf4ba57f4bdbf

**Comment by badlogic** [comment]
Created an alternative implementation that doesn't require a scratch buffer. Easier to port.

**Comment by NathanSweet** [comment]
Editor should not allow a linked mesh to be the parent mesh for another linked mesh. Fixed in 3.7.92, on project load:
![](http://n4te.com/x/6146-GGJN.png)

**Comment by HaraldCsaszar** [comment]
User `Fainder` on the forum has reported some repro steps:
"It seems if I duplicate a skin that has some new meshes and some linked meshes, the duplicate creates linked meshes that refer to other linked meshes."
http://esotericsoftware.com/forum/Error-reading-Skeleton-JSON-file-in-Unity-11567

**Comment by NathanSweet** [comment]
Fixed in 3.7.92!
